tests/test_apply.py

```diff
@@ -1,16 +1,179 @@
-import tempfile
 import os
-import boto3
-import uuid
+import random
 import subprocess
+import tempfile
+import uuid
 from typing import Dict
 
+import boto3
+
 THIS_PATH = os.path.abspath(os.path.dirname(__file__))
 ROOT_PATH = os.path.join(THIS_PATH, "..")
 TFLOCAL_BIN = os.path.join(ROOT_PATH, "bin", "tflocal")
 LOCALSTACK_ENDPOINT = "http://localhost:4566"
 
 
+def test_access_key_override_fallback(monkeypatch):
+    access_key = f"{mock_access_key()}"
+    monkeypatch.setenv("AWS_ACCESS_KEY_ID", access_key)
+    bucket_name = f"{short_uid()}"
+    config = """
+    provider "aws" {
+      region = "eu-west-1"
+    }
+    resource "aws_s3_bucket" "test_bucket" {
+      bucket = "%s"
+    }""" % (bucket_name)
+    deploy_tf_script(config)
+
+    s3 = client("s3", region_name="eu-west-1")
+    s3_buckets = s3.list_buckets().get("Buckets")
+    s3_bucket_names = [s["Name"] for s in s3_buckets]
+
+    assert bucket_name in s3_bucket_names
+
+    s3 = client("s3", region_name="eu-west-1", aws_access_key_id=access_key)
+    s3_buckets = s3.list_buckets().get("Buckets")
+    s3_bucket_names = [s["Name"] for s in s3_buckets]
+
+    assert bucket_name not in s3_bucket_names
+
+
+def test_access_key_override_by_env_var(monkeypatch):
```

Comment:
nit: this test is quite similar to `test_access_key_override_fallback` above - perhaps we could consider combining into a single function, and parametrizing the test via `@pytest.mark.parametrize("customize_access_key", [True, False])`

(not critical at all though, just an idea for future iterations, no need to update the PR at this point)

Reply:
The cases are slightly different as the assertion is switching over so just introducing the parameterization is not enough here. In the next commit I'll add some sub-functions so we reduce the code repetition.